com.vrcfury.vrcfury/Editor/VF/Feature/CombineMeshesBuilder.cs

```diff
@@ -0,0 +1,543 @@
+
+namespace VF.Feature
+{
+    public class CombineMeshesBuilder : FeatureBuilder<CombineMeshes>
+    {
+        [VFAutowired] private readonly ObjectMoveService mover;
+
+        private static string[] preferredBaseMeshNames = new string[] { "body", "main" };
+
+        public override string GetEditorTitle()
+        {
+            return "Combine Meshes";
+        }
+
+        public override VisualElement CreateEditor(SerializedProperty prop)
+        {
+            var content = new VisualElement();
+            content.Add(VRCFuryEditorUtils.Info(
+                "This feature will combine all similar meshes that are not animated to be toggled into one, " +
+                " saving extra draw calls!"
+            ));
+            content.Add(new VisualElement { style = { paddingTop = 10 } });
+            content.Add(VRCFuryEditorUtils.Prop(prop.FindPropertyRelative("keepFaceMeshSeparate"), "Keep Face Mesh separate"));
+            content.Add(new VisualElement { style = { paddingTop = 10 } });
+            return content;
+        }
+
+        public override bool AvailableOnProps()
+        {
+            return false;
+        }
+
+        [FeatureBuilderAction(FeatureOrder.BlendshapeOptimizer)]
+        public void Apply()
+        {
+            var changes = CalculateMeshChanges();
+            foreach (var group in changes.Groups)
+            {
+                var skins = group.Value;
+                var basis = skins[0];
+                VFGameObject basisObject = basis.gameObject;
+                var meshesToAdd = skins.Skip(1).ToList();
+                basis.sharedMesh = mutableManager.MakeMutable(basis.sharedMesh, basis.owner()); ;
+                VRCFuryEditorUtils.MarkDirty(basis);
+
+                var combinable = new CombinableMesh(basis);
+                foreach (var mesh in meshesToAdd)
+                {
+                    combinable.AddMesh(mesh);
+                    VFGameObject obj = mesh.gameObject;
+                    mover.DirectRewrite(obj, basisObject);
+                    obj.Destroy();
+                }
+                combinable.Combine();
+            }
+            foreach (var skin in changes.Removed)
+            {
+                VFGameObject obj = skin.gameObject;
+                obj.Destroy();
+            }
+        }
+
+        private struct MeshChanges
+        {
+            public Dictionary<SkinnedMeshRendererGroup, List<SkinnedMeshRenderer>> Groups { get; set; }
+            public List<SkinnedMeshRenderer> Toggled { get; set; }
+            public List<SkinnedMeshRenderer> Removed { get; set; }
+            public List<SkinnedMeshRenderer> NotCombinable { get; set; }
+        }
+
+        private MeshChanges CalculateMeshChanges()
+        {
+            var task = new MeshChanges();
+            task.Groups = new Dictionary<SkinnedMeshRendererGroup, List<SkinnedMeshRenderer>>();
+            task.Removed = new List<SkinnedMeshRenderer>();
```

Comment:
Would it be safe to also combine non-skinned meshes into the main skinned renderer? 🤔

Reply:
Depends on where they are. If they are a child somewhere in the armature, you'd need to weight paint the vertices of those meshes to the bone, which isn't hard to do and I actually mentioned it in a previous comment that it's possible. But I didn't want to spend even more time on this when it's not clear that it's getting merged :) If this gets in I'll work on that next.

Reply:
In theory we could also detect parent constraints and merge them too.